### ISSUE TITLE
When playing multiplayer, destroying a unit by damage crashes one client

### ISSUE THREAD

**Issue by kamarapala** [calmackenzie/VG1819#178]
TL;DR:
- Unit gets destroyed because of a damaging ability received over the network
- Game crashes because game is trying to call update() on a LerpController component which was deleted but is still getting called from K_ComponentManager's m_toUpdate because it was hiding in m_toAddToUpdate which added it there after it was deleted.
- One possible solution to this would be to add `m_toAddToUpdate.remove(p_toRemove)` to K_ComponentManager::removeFromUpdate
- Finally figured out why this only happens to the client that receives the ability over the network: on a local client use of an ability, the action to use the ability originates in InputManager, which updates before the component manager and the game object manager. 
On receiving an ability over the network, it originates from a component update(), in the middle of the component manager update, causing the hiding LerpController to be added to m_toUpdate on the next frame, after the associated game object has been deleted

Although the real issue is that the networking stuff is updating from inside a component and should probably be moved to K_Instance, @calmackenzie you should see if it's worth also clearing m_toAddToUpdate when components are being deleted.

Note: I was adding to this while actively trying to solve this issue, so everything below this may have unnecessary information. The TL;DR should have most of the necessary information about what the issue is and why it was being caused.

After using an ability that ends up destroying a unit by damage, the client that receives the used ability over the network crashes after the unit is destroyed, while the other client that originally used the ability is fine.

Call stack:
![](https://i.imgur.com/6Y4GxgI.png)

The issues is that K_ComponentManager is trying to call update() on a deleted component. 
The deleted component is the LerpController attached to the foreground health bar game object created in a UnitHealthBar component which is attached to the unit game object that's being destroyed.

What's happening is that when the damaging ability is used, eventually ChangeAttributeNode::effect is called and inside, CheckHPNode::effect gets called and UnitHealthBar::updateBar also gets called immediately after. 
CheckHPNode::effect determines that the unit should be destroyed by damage and begins the process of deleting the game object (where the deletion is finally queued with a call to K_GameObjectManager::destroyGameObjectWithChild in InitiativeTracker::removeUnit).

However, as UnitHealthBar::updateBar is called immediately after CheckHPNode::effect, UnitHealthBar::updateBar enables the previously disabled LerpController through a call to LerpController::scaleLerp. 
This puts the LerpController into a queue (m_toAddToUpdate list) to add the component into K_ComponentManager's m_toUpdate unordered_set.

This leads to the issue, the unit game object finally gets destroyed, destroying all attached components with calls to K_ComponentManager::destroyComponentImmediate. However, this function can only remove components that may exist in the m_toUpdate and m_toAddToStart unordered_sets, while a pointer to the LerpController is hiding in m_toAddToUpdate, which gets added to m_toUpdate on the next frame, but by which point the pointer has been deleted.

### Possible Solution
A possible solution to this would be to add `m_toAddToUpdate.remove(p_toRemove)` to K_ComponentManager::removeFromUpdate, which might be a good idea if this issue could possibly happen again elsewhere in case this could happen in another situation.. 

### Why The Issue Occurs on Multiplayer
The reason why this happens only to the client receiving the ability over the network is because for the client that originally uses the ability, the starting point is in the InputManager (specifically, the input leads to ActionSelect::onClick), which, in the case of using an ability that destroys a unit, eventually calls all the way to deleting the unit game object and deleting all the components. The difference here being that since the InputManager updates before K_ComponentManager::updateComponents, which also updates before K_GameObjectManager::deleteQueuedObjects, the LerpController that is hiding in m_toAddToUpdate manages to get added to m_toUpdate before the game object and components are deleted.

For the client that receives the ability over the network, the starting point is when they receive the Ability Packet in ClientGame::update, which is called inside of a K_Component::update in the middle of K_ComponentManager::updateComponents, which is immediately followed by a call to K_GameObjectManager::deleteQueuedObjects, causing the hiding LerpController to be added to m_toUpdate on the next frame after the unit game object and related components are deleted, creating the issue.
